src/RefinedAst.hs

```diff
@@ -8,32 +8,51 @@
 {-# Language DeriveAnyClass #-}
 module RefinedAst where
 import Data.Text          (Text, pack, unpack)
+import Data.List (filter)
 import GHC.Generics
 import Data.Map.Strict    (Map)
 import Data.List.NonEmpty hiding (fromList)
 import Data.ByteString       (ByteString)
 
-import Syntax hiding (Invariants)
+import Syntax
 import Data.Aeson
 import Data.Aeson.Types
 import Data.Vector (fromList)
 
 -- AST post typechecking
-data Behaviour =
-  Invariants [Exp Bool] [Id]
-  | Behaviour
-    {_name :: Id,
-     _mode :: Mode,
-     _creation :: Bool,
-     _contract :: Id,
-     _interface :: Interface,
-     _preconditions :: Exp Bool,
-     _postconditions :: Exp Bool,
-     _contracts :: [Id], -- can maybe be removed; should be equivalent to Map.keys(_stateupdates)
-     _stateUpdates :: Map Id [Either StorageLocation StorageUpdate],
-     _returns :: Maybe ReturnExp
-    }
-
+data Claim = B Behaviour | I Invariant
+
+data Invariant = Invariant Id (Exp Bool)
+data Behaviour = Behaviour
+  {_name :: Id,
+   _mode :: Mode,
+   _creation :: Bool,
+   _contract :: Id,
+   _interface :: Interface,
+   _preconditions :: Exp Bool,
+   _postconditions :: Exp Bool,
+   _contracts :: [Id], -- can maybe be removed; should be equivalent to Map.keys(_stateupdates)
+   _stateUpdates :: Map Id [Either StorageLocation StorageUpdate],
+   _returns :: Maybe ReturnExp
+  }
+
+catInvs :: [Claim] -> [Invariant]
+catInvs = fmap unwrapI . Data.List.filter isI
```

Comment:
```suggestion
catInvs = fmap unwrapI . filter isI
```

Reply:
another way.
```
 catInvs :: [Claim] -> [Invariant] 
catInvs [] = []
catInvs ((I i):claims) = i:(catInvs claims)
catInvs ((B _):claims) = catInvs claims
```
feel free to use whichever you prefer

Reply:
ah nice, total functions!!! thx :pray: